README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# recurpkg

<!-- badges: start -->

<!-- badges: end -->

The goal of recurpkg is to calculate a sequence vector defined as
following: \(X_n\) = \(X_{n-1}\) +|\(X_{n-2} - X_{n-3\)}|/2

That is, element n is the sum of element n−1 and the absolute value of
the difference between elements n−2 and n−3 divided by two where each
element is a number. For example, if we let \(X_1\)=3, \(X_2\)=1, and
\(X_3\)=10, then \(X_4\) is \(X_4\) = 10+ |1−3|/2 =11.

If the input is a vector, then the initial three numeric values will be
taken as initial values. If the input is a dataframe, then the dataframe
should has 4 columns, while the first three columns are the initial
values of the sequence and the last columns will be the nth element of
the sequence. The output will be a plot with x axis as the sequence
position n, y axis as the sequence value.

## Installation

"This package is only available by permission of the author at

install.packages(“<https://github.com/fangfeiliu/recurpkg.git>”)

``` r
install.packages("recurpkg")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(recurpkg)
## basic example code
## vector calculation
## vector length is equal to three
recur_vec(x = c(2, 4, 3), n = 3)
#> [1] 3
```

``` r
## vector length is greater than 3
recur_vec(x = c(2, 4,3,4), n = 3)
#> [1] 3
```

``` r
## vector length is equal to three and the sequence number is greater than 3
recur_vec(x = c(2, 4, 3), n = 5)
#> [1] 4.5
```

``` r
## sequence number is not an integar
recur_vec(x = c(2, 4, 3), n = 4.5)
#> [1] "the sequence number indicator must be an integer"
```

``` r
## vector contains both numeric and character values
recur_vec(x = c(2,'a',4,'b', 3), n = 4)
#> [1] 4
```

``` r
#dataset example
my_data <- tibble::tribble(
  ~x, ~y, ~z, ~n,
  2,4,3,3,
  2,4,3,5,
  2,4,3,7,
  2,4,3,10,
  2,4,3,15,
  2,4,3,20,
  2,4,3,25
)

recur_dataset(my_data)
```

<img src="man/figures/README-unnamed-chunk-6-1.png" width="100%" />


NEWS.md
# recurpkg 0.0.0.9000

## first created date July 24th 2020. version 0.0.0.9000
* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(recurpkg)

test_check("recurpkg")


R/recur.R
## funciton for vector
recur_vec <- function(x,n) {
  if(n <= 0) return(" Position numner must be greater than zero")
  if (length(x) < 3 && n >= length(x)) return ("Please provide a valid vector (length > 3)")
  else if(length(x) <= 3 && n <= length(x)) return (x[n])
  d = c(seq(1:n))
  d[1] = x[1]
  d[2] = x[2]
  d[3] = x[3]
  for (i in 4:n) {
    d[i] = d[i-1]+abs(d[i-2]-d[i-3])/2
  }
  return(d[n])}


#function for dataset and graph
recur_dataset <- function(x) {
  position_value = c(seq(1:nrow(x)))
  seq_number = c(seq(1:nrow(x)))
  for (i in 1:nrow(x)) {
    if(length(x[i,]) != 4) return("Incorrect column number in the dataset")
    vec_value <- as.numeric(x[i,][1:3])
    position_value[i]<- as.numeric(x[i,][4])
    seq_number[i] <- recur_vec(vec_value,position_value[i])
  }
  graph_dt = tibble(position_value,seq_number)

  ggplot(graph_dt, aes(x=position_value,y = seq_number)) +
    geom_line()+
    ggtitle(paste("My Sequences: c(",paste(as.character(seq_number),collapse=", "),")",collapse ="")) +
    xlab("n")+
    ylab("output") -> q
  return(q)
}


R/recur_vec.R
#' function for vector recursive sequence calculation
#' Recurpackage recur_vec package
#' The input should have one vector and a sequence number indicator n ,
#' The vector should have more than one elements while the sequence number indicator n
#' should be a positive integar (>n)
#' The output is the final nth element of the sequence to calculate defined as following
#'  X_n = X_(n-1) + |X_(n-2) -X(n-3)|/2
#' @param x
#' x should be a vector with more than one elements. However, the function will only take
#' the first three elements of the vector as the initial value as the recursive sequence
#' initial value input
#' @param n
#' n should be a positive integer number
#' @return
#' The output is the final nth element of the sequence to calculate defined as following
#'  X_n = X_(n-1) + |X_(n-2) -X(n-3)|/2
#' @export
#' recur_vec
#' @examples
#' recur_vec(x = c(2, 4, 3), n = 3)
#' recur_vec(x = c(2, 4, 3), n = 4)
#' recur_vec(x = c(2, 4, 3,5), n = 3)

recur_vec <- function(x,n) {
  x = as.numeric(x)
  x = x[!is.na(x)]
  if(length(x) > 3) x = x[1:3]
  if(as.integer(n) != n) return ("the sequence number indicator must be an integer")
  if(n <= 0) return(" Position numner must be greater than zero")
  if (length(x) < 3 && n >= length(x)) return ("Please provide a valid vector (length > 3)")
  else if(length(x) <= 3 && n <= length(x)) return (x[n])
  d = c(seq(1:n))
  d[1] = x[1]
  d[2] = x[2]
  d[3] = x[3]
  for (i in 4:n) {
    d[i] = d[i-1]+abs(d[i-2]-d[i-3])/2
  }
  return(d[n])}


R/recur_dataset.R
#' function for dataset and graph
#' Recurpackage recur_dataset package
#' The input data set should be 4 columns,
#' The first three columns are the values of the three
#' numbers to be input in order to calculate the recursive sequence define by
#'  X_n = X_(n-1) + |X_(n-2) -X(n-3)|/2  and the fourth column
#' is the positive integer n for the sequence to be generated.
#' The output is expected to be a line plot with X axis as n and Y axis as sequence numbers
#' @param x
#' 4 columns wide dataset with the first three columns are the values of the three
#' numbers to be input
#' @return
#' return a line plot with x axis as the sequence length n and Y axis as the sequence number
#' vector output
#' @export
#' recur_dataset
#'
#' @examples
#' my_data <- tibble::tribble(
#' ~x, ~y, ~z, ~n,
#' 2,4,3,3,
#' 2,4,3,5,
#' 2,4,3,7,
#' 2,4,3,10,
#' 2,4,3,15,
#' 2,4,3,20,
#' 2,4,3,25
#' )
#'
#' recur_dataset(my_data)
recur_dataset <- function(x) {
  position_value = c(seq(1:nrow(x)))
  seq_number = c(seq(1:nrow(x)))
  for (i in 1:nrow(x)) {
    if(length(x[i,]) != 4) return("Incorrect column number in the dataset")
    vec_value <- as.numeric(x[i,][1:3])
    position_value[i]<- as.numeric(x[i,][4])
    seq_number[i] <- recur_vec(vec_value,position_value[i])
  }
  graph_dt = tibble(position_value,seq_number)

  ggplot(graph_dt, aes(x=position_value,y = seq_number)) +
    geom_line()+
    ggtitle(paste("My Sequences: c(",paste(as.character(seq_number),collapse=", "),")",collapse ="")) +
    xlab("n")+
    ylab("output") -> q
  return(q)
}
